### PR TITLE
neovim: Add coc support

### PR DESCRIPTION
**Pull request by t4ccer**
### Description

Adds options to enable [coc](https://github.com/neoclide/coc.nvim) and provide content of `coc-settings.json`

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/neovim.nix [hunk 1]
```diff
@@ -241,9 +274,12 @@ in {
 
     home.packages = [ cfg.finalPackage ];
 
-    xdg.configFile = mkIf (neovimConfig.neovimRcContent != "") {
-      "nvim/init.vim".text = neovimConfig.neovimRcContent;
+    xdg.configFile."nvim/init.vim" = mkIf (neovimConfig.neovimRcContent != "") {
+      text = neovimConfig.neovimRcContent;
```
Comment (kalbasit) [inline modules/programs/neovim.nix:278]: Does this end up creating the `init.vim` file even if the contents were supposed to be empty?

Comment (t4ccer) [reply modules/programs/neovim.nix:278]: I have just tested it and it won't create `init.vim` when condition is `false`. However it actually does create file due to #1907

File: modules/programs/neovim.nix [hunk 2]
```diff
@@ -217,6 +217,38 @@ in {
           This option is mutually exclusive with <varname>configure</varname>.
         '';
       };
+
+      coc = {
+        enable = mkEnableOption "Coc";
+
+        extraConfig = mkOption {
```
Comment (ambroisie) [inline modules/programs/neovim.nix:224]: Why not use a `settings` option, using `formats.json`?

Comment (berbiche) [reply modules/programs/neovim.nix:224]: Yes, `extraConfig` will not be accepted if the configuration can use structured types.

Comment (berbiche) [reply modules/programs/neovim.nix:224]: As an example: https://github.com/nix-community/home-manager/blob/db00b39a9abec04245486a01b236b8d9734c9ad0/modules/services/hound.nix.

This module uses `pkgs.formats.json`.

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks!

Please use a structured type instead of `extraConfig`.

See this comment: https://github.com/nix-community/home-manager/pull/2154#discussion_r663206462

**Comment by t4ccer** [comment]
Ok, I didn't think about using `formats.json`. I will reimplement it that way

**Comment by berbiche** [comment]
Friendly ping @t4ccer

**Comment by t4ccer** [comment]
Sorry for the delay, have been afk for some time

File: modules/programs/neovim.nix [hunk 3]
```diff
@@ -217,6 +219,38 @@ in {
...
+              "suggest.enablePreselect":false,
+              "suggest.disableKind": true,
+              "languageserver": {
+                "haskell": {
+                  "command": "haskell-language-server-wrapper",
+                  "args": ["--lsp"],
+                  "rootPatterns": ["*.cabal", "stack.yaml", "cabal.project", "package.yaml", "hie.yaml"],
+                  "filetypes": ["haskell", "lhaskell"]
+                }
+              }
```
Comment (berbiche) [inline modules/programs/neovim.nix:242]: Can you change the code to use the nix syntax since we are no longer using `types.lines`

**Review by berbiche — CHANGES_REQUESTED** [review]

File: modules/programs/neovim.nix (hunk 3 above)
Comment (t4ccer) [reply modules/programs/neovim.nix]: Changed

File: modules/programs/neovim.nix [hunk 4]
```diff
@@ -226,22 +226,26 @@ in {
...
-                "haskell": {
-                  "command": "haskell-language-server-wrapper",
-                  "args": ["--lsp"],
-                  "rootPatterns": ["*.cabal", "stack.yaml", "cabal.project", "package.yaml", "hie.yaml"],
-                  "filetypes": ["haskell", "lhaskell"]
-                }
-              }
-            }
-          '';
+          example = {
```
Comment (berbiche) [inline modules/programs/neovim.nix:229]:
```suggestion
          example = literalExample ''
            {
```

Sorry, last thing! You need to keep the literalExample when using attribute sets or lists in examples.

**Review by berbiche — CHANGES_REQUESTED** [review]

File: modules/programs/neovim.nix (hunk 4 above)
Comment (t4ccer) [reply modules/programs/neovim.nix]: Done

File: modules/programs/neovim.nix [hunk 5]
```diff
@@ -225,7 +265,8 @@ in {
       inherit (cfg)
         extraPython3Packages withPython3 withNodeJs withRuby viAlias vimAlias;
       configure = cfg.configure // moduleConfigure;
-      plugins = cfg.plugins;
+      plugins = cfg.plugins
+        ++ (if cfg.coc.enable then [ pkgs.vimPlugins.coc-nvim ] else [ ]);
```
Comment (berbiche) [inline modules/programs/neovim.nix:269]:
```suggestion
        ++ optionals cfg.coc.enable [ pkgs.vimPlugins.coc-nvim ];
```

**Review by berbiche — APPROVED** [review]
Thanks! I'll merge this when I'll be at my computer.

**Pull request merged**